src/auth.rs

```diff
@@ -0,0 +1,247 @@
+//! A wrapper for performing Remote Framebuffer authentication.
+
+use anyhow::{bail, Context, Result};
+
+use futures::{SinkExt, StreamExt};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+/// What kind of authentication to use for an RFB connection.
+pub enum RfbAuthentication {
+    /// A null authentication. It does not perform the initial handshake, so it relies on the rest
+    /// of the connection to pass through the data as-is without parsing.
+    Null,
+
+    /// An authentication that parses the initial ProtocolVersion and Security handshakes, and
+    /// passes them as-is to the peer, without acting on it. This leaves the stream in a state
+    /// where the ClientInit handshake is expected to appear next, followed by a stream of normal
+    /// RFB messages can appear.
+    Passthrough,
+}
+
+/// A way of authenticating an RFB connection.
+pub async fn authenticate<SocketStream, WebSocketStream>(
+    authentication: &RfbAuthentication,
+    stream: &mut SocketStream,
+    ws_stream: &mut WebSocketStream,
+) -> Result<()>
+where
+    SocketStream: tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin + Send,
+    WebSocketStream: futures::Sink<
```

Comment:
this is fun!

Reply:
it also took me a good while to figure out how to do properly. had to unlearn all those years of OOP indoctrination, because my first instinct was to build a `trait` instead of an `enum`, but then the generic types wouldn't match because I didn't know from the place i needed to instantiate the trait object what kind of types I needed, and I couldn't quite figure out how to do type erasure in Rust properly. and then Rust didn't quite like that I had more than one `dyn` trait.

but yes, learning new things _is_ fun :D